nexus/src/sagas.rs

```diff
@@ -927,30 +1039,120 @@ const MAX_CONCURRENT_REGION_REQUESTS: usize = 3;
 
 async fn sdc_regions_ensure(
     sagactx: ActionContext<SagaDiskCreate>,
-) -> Result<(), ActionError> {
+) -> Result<String, ActionError> {
     let log = sagactx.user_data().log();
     let datasets_and_regions = sagactx
         .lookup::<Vec<(db::model::Dataset, db::model::Region)>>(
             "datasets_and_regions",
         )?;
+
     let request_count = datasets_and_regions.len();
-    futures::stream::iter(datasets_and_regions)
+
+    // Allocate regions, and additionally return the dataset that the region was
+    // allocated in.
+    let datasets_and_regions: Vec<(
+        db::model::Dataset,
+        crucible_agent_client::types::Region,
+    )> = futures::stream::iter(datasets_and_regions)
         .map(|(dataset, region)| async move {
-            ensure_region_in_dataset(log, &dataset, &region).await
+            match ensure_region_in_dataset(log, &dataset, &region).await {
+                Ok(result) => Ok((dataset, result)),
+                Err(e) => Err(e),
+            }
         })
         // Execute the allocation requests concurrently.
         .buffer_unordered(std::cmp::min(
             request_count,
             MAX_CONCURRENT_REGION_REQUESTS,
         ))
-        .collect::<Vec<Result<_, _>>>()
+        .collect::<Vec<
+            Result<
+                (db::model::Dataset, crucible_agent_client::types::Region),
+                Error,
+            >,
+        >>()
         .await
         .into_iter()
-        .collect::<Result<Vec<_>, _>>()
+        .collect::<Result<
+            Vec<(db::model::Dataset, crucible_agent_client::types::Region)>,
+            Error,
+        >>()
         .map_err(ActionError::action_failed)?;
 
-    // TODO: Region has a port value, we could store this in the DB?
-    Ok(())
+    // Assert each region has the same block size, otherwise Volume creation
+    // will fail.
+    let all_region_have_same_block_size = datasets_and_regions
+        .windows(2)
+        .all(|w| w[0].1.block_size == w[1].1.block_size);
+
+    if !all_region_have_same_block_size {
+        return Err(ActionError::new_subsaga(
+            // XXX wrong error type?
```

Comment:
Similar to my comments re: TODOs above - the XXX comments here should probably be fleshed out a bit? It's okay to document shortcomings, but I'm trying to get a better picture of "comments from James to James before merging" vs "comments for everyone to consume".

Reply:
Removed some XXX in db559795, now I don't see any if I diff against the merge base. Will follow up on the TODOs.